### DISCUSSION TITLE
How to use local interrupt?

### DISCUSSION THREAD

**Discussion by Usutatsu** [polarfire-soc/polarfire-soc-documentation#238] [How to] [answered]
I use PolarFire SoC Icicle kit and I want to evaluate some program.
Thus, I would like to use local interrupt from FPGA fabric not using CLINT.

Specifically, I want to enable any of the local interrupts 0~47 to interrupt when SW1 on the board is pressed.
How should the hardware be implemented and programmed for such an interrupt in U54_1.
I would like to implement in bare-metal.

I think that the signal line for interrupts should use MSS_INT_F2M[63:0], but is this correct?
Also, if there is anything that would be helpful for the program, please let me know.

**Comment by hughbreslin** [comment, marked as answer]
Hi @Usutatsu I'm a bit confused by "not using CLINT" - the CLINT is used for local interrupt generation so you can't really bypass it for this use case. You could use the PLIC or a GPIO for interrupts instead of the CLINT but these are global interrupts not local. 

You are correct in thinking you can use the MSS_INT_FTM signals - for an example of this have a look at this section of our [GPIO interrupt example](https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/blob/main/driver-examples/mss/mss-gpio/mpfs-gpio-interrupt/src/application/hart1/u54_1.c#L147) which shows the MSS_INT_FTM (in this case called FABRIC_F2H_...) driving an interrupt. You will need to configure the PLIC for this to work.

If you want to use the local interrupts on the hart (which does use the CLINT) you can use the `__enable_local_irq` function that is shown [here](https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/blob/main/driver-examples/mss/mss-gpio/mpfs-gpio-interrupt/src/platform/mpfs_hal/common/mss_util.c#L75). Just note that not all interrupts are available on all harts, MSS_INT_F2M[31:0] are available for the U54 harts, while MSS_INT_F2M[63:32] are available for the E51, this information is available in table 6-1 of the [PolarFire SoC TRM](https://www.microsemi.com/document-portal/doc_view/1245725-polarfire-soc-fpga-mss-technical-reference-manual).

For examples of how to connect these interrupts have a look at the [Icicle Kit Reference Design](https://github.com/polarfire-soc/icicle-kit-reference-design).

**Reply by hughbreslin** [reply]
For more information on the interrupt connections have a look at figure 6-1 of the TRM linked above :)

**Reply by Usutatsu** [reply]
Thank for your comments. I understood how to use local interrupt.
I wired the interrupt signal. Then I also created the program.
As a result, we were able to confirm that an interrupt is generated when the switch is pressed. However, once the switch is pressed, the interrupt is repeatedly triggered. Is there a way to modify the logic or program so that pressing the switch once will cause the interrupt to be triggered only once?

**Reply by hughbreslin** [reply]
My understanding is that the CLINT interrupt are level triggered, so the switch staying pressed causes the core to jump to the handler continuously.

I have two suggestions:

1. Implement a pulse generator in the fabric so that once the switch is pressed it will generate a single pulse to the interrupt line and won't generate another pulse until the switch input has gone low and then high again. 
2. The other option is using one of the MSS GPIO blocks as the interrupt controller (like what is done in the GPIO interrupt example) - section 4.12.9.2 of the TRM describes the different interrupt modes the GPIO block can be configured in and one of them is edge triggered instead of level triggered. Using the GPIO will have higher latency than using the CLINT but probably not by much depending on how you configure it. 

If you go the GPIO route make sure to set up the `GPIO_INTERRUPT_FAB_CR` register correctly to route the interrupt signals - its set up through the MSS configurator and theres a description [here](https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/blob/main/driver-examples/mss/mss-gpio/mpfs-gpio-interrupt/src/platform/drivers/mss/mss_gpio/mss_gpio.h#L127).

**Reply by Usutatsu** [reply]
Thank you for your kind instruction.
My questions have been answered.

I have another question, in the latest reference design in the polarfire soc icicle kit, I think the global clock of the FPGA fabric is a 160Mhz oscillator.
I would like to set the global clock of the FPGA to 80Mhz. How should I modify my design at this time?
If you have any knowledge about this, I would like to know.

**Comment by hughbreslin** [comment]
The Icicle Kit Reference Design is using a 50MHz on board oscillator as the source clock now - this goes into a CCC (Clock Conditioning Circuitry) block to generate the fabric clocks (this can be found in the "CLOCKS_AND_RESETS" SmartDesign):

![image](https://user-images.githubusercontent.com/44862881/201054558-c9c56388-f8e4-4fc0-a4e9-8667a87f8c03.png)

To update the clock frequency if you have a different source clock you need to update the configuration of the CCC:

![image](https://user-images.githubusercontent.com/44862881/201054929-e07ae0ae-867a-48d6-a486-9ff86e3ca425.png)

**Reply by Usutatsu** [reply]
Thank you for your insightful comments.

I have an additional question, can I use the oscillator input to the 125MHZ MSS instead of the 50MHZ on-chip oscillator?
Additionally, can I change the feedback mode of the CCC to external feedback?
